P2047-alloc-aware-optional.html

```diff
@@ -38,128 +38,70 @@
-  memory, such as <tt>tuple</tt>, follow the AA rules when they hold one more
-  more AA elements. (A special case is <tt>pair</tt>, which is not actually AA
-  but effectively follows the rules through special handling
-  in <em>uses-allocator construction</em>.)
-</p>
+- replaced <tt>alloc_optional</tt> with additional overloads of <tt>make_optional</tt> that 
+take an <tt>allocator_arg_t</tt> : it was observed that <tt>alloc_optional</tt> doesn't actually 
+do any allocation, and that the name is not appropriate
+- fixed the order of arguments in the <i>Returns</i> clause of what used to be <tt>alloc_optional</tt>
+- adjusted for LWG 2833
+- added discussion on deduction guides
+- removed mentions of using a type alias since the type alias approach has proven to be problematic
+- modified the specification of pmr::optional to cover both AA and non AA types
+- added CTAD specification
 
-<p>
-  The current definition of <tt>std::optional</tt> does not follow the rules
-  for an AA type, even when holding an AA value. This limitation
-  makes <tt>optional</tt> unsuitable for storage in an AA container when
-  memory allocation customization is needed.
-</p>
 
 <p>
-  In this paper, we propose a new, allocator-aware
-  <tt>std::pmr::optional</tt> usable as a container element type and in any
-  other context where allocator propagation is expected. This new type would
-  not replace the current <tt>std::optional</tt>, but would resolve to
-  <tt>std::optional</tt> when the value type is not AA. By automatically
-  handling the distinction between AA and non-AA value types,
-  <tt>pmr::optional</tt> can easily be used in generic code where it is not
-  known in advance if it would be storing an AA type or not.
+  This is a proposal for an allocator-aware optional type. For users
+  that use custom allocators, such a type is currently missing;
+  pair and tuple support allocators, but an optional type is also
+  deemed necessary, because some such users use optional types
+  as container elements. Therefore, an allocator-aware optional
+  type is necessary for the usual scoped-allocator use cases to work.
+<br/>  
+  This is a complete proposal. It has not been seen by LEWGI or LEWG yet.
 </p>
 
+<h2> Motivation</h2>
 <p>
-  This is a complete proposal with formal wording.
-  It has not yet been discussed in LEWGI or LEWG.
+  Current optional type does not retain the element's allocator when it is
+  disengaged. This is a problem when using pmr containers because there is
+  no way of ensuring that a pmr::vector&lt;optional&gt; has all of its
+  elements using the same allocator - if an optional object gets disengaged, 
+  the pmr container it belongs to can not ensure the same allocator will 
+  be used if the optional object is re-engaged.
+  We propose an optional type which ensures that all of its elements use the same allocator.  
+  The allocator is not part of the type in the sense that it's 
+  not a template parameter.
```

Comment:
While this is all correct as such, I wonder whether this explanation should mention that this is done in order for scoped allocation to work. That is, if we have a pmr::vector<pmr::optional<pmr::string>>>, then the polymorphic allocator given to the vector will also allocate the strings. I know you probably mean to say that with "all elements use the same allocator", but I think those are two slightly different things: 1) the container makes a good effort to have all its elements use the same allocator 2) but on the other hand, it's specifically the container's allocator that is used via scoped allocation for the pmr::optional, and by extension for the pmr::string in it. 

I suppose one way to explain the difference is that "all elements use the same allocator" would be true if they'd all use a default-constructed polymorphic_allocator (which gets the default_memory_resource), but scoped allocation means that all elements use the container's allocator.

And if we want to be a bit pedantic, the elements might not use the same allocator, but they will use the same memory_resource, and that's the memory_resource that the container uses.

And finally, if we want to go for a complex but perhaps more explanatory example, make it a 
pmr::vector<pmr::optional<pmr::vector<pmr::optional<pmr::string>>>>. Without allocator support in pmr::optional, the memory_resource of the topmost vector wouldn't be scoped through all of those pmr-things, but with the allocator support, it is.

Reply:
it would appear i have misplaced the improvements to this paper that cover this. I've reinstated them. Let me know if you think there is still room for improvement